File: backend/src/main/java/com/shiptrackpro/backend/auth/service/AuthService.java
package com.shiptrackpro.backend.auth.service;

import com.shiptrackpro.backend.auth.dto.*;
import com.shiptrackpro.backend.auth.entity.RefreshToken;
import com.shiptrackpro.backend.auth.repository.RefreshTokenRepository;
import com.shiptrackpro.backend.common.config.security.CustomUserDetails;
import com.shiptrackpro.backend.common.config.security.JwtService;
import com.shiptrackpro.backend.user.entity.Role;
import com.shiptrackpro.backend.user.entity.RoleName;
import com.shiptrackpro.backend.user.entity.User;
import com.shiptrackpro.backend.user.entity.UserStatus;
import com.shiptrackpro.backend.user.repository.RoleRepository;
import com.shiptrackpro.backend.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.ZonedDateTime;
import java.util.Base64;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final JwtService jwtService;
    private final RefreshTokenRepository refreshTokenRepository;
    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;

    @Value("${jwt.access.expiration:900000}")
    private long jwtExpirationMs;

    @Value("${jwt.refresh.expiration:604800000}")
    private long refreshTokenDurationMs;

    @Transactional
    public LoginResponse login(LoginRequest request) {
        Authentication authentication;
        try {
            authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(request.getEmail().toLowerCase(), request.getPassword()));
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid email or password");
        }

        SecurityContextHolder.getContext().setAuthentication(authentication);
        CustomUserDetails userDetails = (CustomUserDetails) authentication.getPrincipal();

        if (userDetails.getUser().getStatus() != UserStatus.ACTIVE) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User is not active");
        }

        String jwt = jwtService.generateToken(userDetails);
        String rawRefreshToken = UUID.randomUUID().toString();

        createRefreshToken(userDetails.getUser(), rawRefreshToken);

        User user = userDetails.getUser();
        user.setLastLoginAt(ZonedDateTime.now());
        userRepository.save(user);

        return LoginResponse.builder()
                .accessToken(jwt)
                .refreshToken(rawRefreshToken)
                .tokenType("Bearer")
                .expiresIn(jwtExpirationMs / 1000)
                .user(mapToCurrentUserResponse(user))
                .build();
    }

    @Transactional
    public LoginResponse register(RegisterRequest request) {
        String email = request.getEmail().toLowerCase();
        if (userRepository.existsByEmail(email)) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email already in use");
        }

        Role customerRole = roleRepository.findByName(RoleName.CUSTOMER)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Default role not found"));

        User user = User.builder()
                .email(email)
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .phone(request.getPhone())
                .status(UserStatus.ACTIVE)
                .roles(java.util.Set.of(customerRole))
                .build();

        userRepository.save(user);

        // Automatically log them in after registration
        LoginRequest loginRequest = new LoginRequest();
        loginRequest.setEmail(email);
        loginRequest.setPassword(request.getPassword());
        return login(loginRequest);
    }

    @Transactional
    public RefreshTokenResponse refresh(RefreshTokenRequest request) {
        String tokenHash = hashToken(request.getRefreshToken());
        
        RefreshToken refreshToken = refreshTokenRepository.findByTokenHash(tokenHash)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid refresh token"));

        if (refreshToken.getRevokedAt() != null || refreshToken.getExpiresAt().isBefore(ZonedDateTime.now())) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid or expired refresh token");
        }

        User user = refreshToken.getUser();
        if (user.getStatus() != UserStatus.ACTIVE) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User is not active");
        }

        CustomUserDetails userDetails = new CustomUserDetails(user);
        String jwt = jwtService.generateToken(userDetails);

        return RefreshTokenResponse.builder()
                .accessToken(jwt)
                .tokenType("Bearer")
                .expiresIn(jwtExpirationMs / 1000)
                .build();
    }

    @Transactional
    public void logout(LogoutRequest request) {
        String tokenHash = hashToken(request.getRefreshToken());
        refreshTokenRepository.findByTokenHash(tokenHash).ifPresent(token -> {
            token.setRevokedAt(ZonedDateTime.now());
            refreshTokenRepository.save(token);
        });
    }

    public CurrentUserResponse getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication == null || !(authentication.getPrincipal() instanceof CustomUserDetails)) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Not authenticated");
        }

        CustomUserDetails userDetails = (CustomUserDetails) authentication.getPrincipal();
        return mapToCurrentUserResponse(userDetails.getUser());
    }

    private void createRefreshToken(User user, String rawToken) {
        RefreshToken refreshToken = RefreshToken.builder()
                .user(user)
                .tokenHash(hashToken(rawToken))
                .expiresAt(ZonedDateTime.now().plusSeconds(refreshTokenDurationMs / 1000))
                .build();
        refreshTokenRepository.save(refreshToken);
    }

    private String hashToken(String token) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] encodedhash = digest.digest(token.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(encodedhash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Failed to hash token", e);
        }
    }

    private CurrentUserResponse mapToCurrentUserResponse(User user) {
        List<String> roles = user.getRoles().stream()
                .map(role -> role.getName().name())
                .collect(Collectors.toList());

        return CurrentUserResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .phone(user.getPhone())
                .status(user.getStatus().name())
                .roles(roles)
                .build();
    }
}

File: frontend/src/services/authService.ts
import { Role, User } from '@/types/domain';
import { apiClient } from './apiClient';

const ROLE_MAP: Record<string, Role> = {
  ADMINISTRATOR: 'Administrator',
  DRIVER: 'Driver',
  CUSTOMER: 'Customer',
  SUPPORT_AGENT: 'SupportAgent',
  BUSINESS_CLIENT: 'BusinessClient',
};

export const authService = {
  async login(email: string, password: string): Promise<{ user: User; accessToken: string }> {
    try {
      const data: any = await apiClient.post('/auth/login', { email, password });
      
      const roleStr = data.user?.role || (data.user?.roles && data.user.roles[0]) || 'CUSTOMER';
      const primaryRole = ROLE_MAP[roleStr.toUpperCase()] || 'Customer';

      const user: User = {
        id: data.user.id,
        name: data.user.name || `${data.user.firstName || ''} ${data.user.lastName || ''}`.trim() || data.user.email,
        email: data.user.email,
        role: primaryRole,
      };

      const token = data.accessToken || data.token;
      localStorage.setItem('accessToken', token);
      if (data.refreshToken) {
        localStorage.setItem('refreshToken', data.refreshToken);
      }
      
      return { user, accessToken: token };
    } catch (e: any) {
      throw new Error(e.message || 'Invalid email or password');
    }
  },

  async register(data: { firstName: string; lastName: string; email: string; password: string; phone?: string }): Promise<any> {
    try {
      const responseData: any = await apiClient.post('/auth/register', data);
      return responseData;
    } catch (e: any) {
      throw new Error(e.message || 'Registration failed');
    }
  },

  async logout(): Promise<void> {
    const refreshToken = localStorage.getItem('refreshToken');
    if (refreshToken) {
      try {
        await apiClient.post('/auth/logout', { refreshToken });
      } catch (e) {
        console.error('Error during logout API call', e);
      }
    }
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
  },

  async getCurrentUser(): Promise<User | null> {
    const token = localStorage.getItem('accessToken');
    if (!token) return null;

    try {
      const data: any = await apiClient.get('/auth/me');
      if (!data) return null;

      const roleStr = data.role || (data.roles && data.roles[0]) || 'CUSTOMER';
      const primaryRole = ROLE_MAP[roleStr.toUpperCase()] || 'Customer';

      return {
        id: data.id,
        name: data.name || `${data.firstName || ''} ${data.lastName || ''}`.trim() || data.email,
        email: data.email,
        role: primaryRole,
      };
    } catch (e) {
      console.error('Error fetching current user', e);
      return null;
    }
  }
};

